joomla/joomla-cms
# Authorization at article level in com_content not always working: Featured and Articles models

### Summary of Changes
Joomla! has a fantastic permissions system that allows fine-grained settings in components, categories till to the article level.
In the major part of cases all these settings are correctly checked and respected but in some circumstancies 'com_content' only checks permissions at component level: article, articles and featured models.

------------------------------------------------------------------------------------------------------------
_[The following problem is now in PR #17735 - BEGIN]_
1. article model:
- `populateState` function authorises at component level and eventually filters published and archived articles.
A 'TODO' in comments seems to me to clearly suggest that a different authorisation was intended.
This PR authorises at article level.

- `getItem` function authorises at component level and eventually filters not-expired and not-future articles.
It seems to me that article level authorisation was simply forgotten.
This PR authorises at article level.

_[Previous problem is now in PR #17735 - END]_

------------------------------------------------------------------------------------------------------------


2. articles model:
- `populateState` function authorises at component level and eventually filters only published articles.
In fact I think that it cannot do otherwisely as we still do not have article ids here.
The only way to authorise at article level here would be to build a SQL query to do all the authorization work, but I think it's impossible, given the complexity of `authorise` code.
This PR postpones authorization and article selection, based on publishing state, to `getItems` function so authorization can be done there at article level.

- `getListQuery` function authorises at component level and eventually filters not-expired and not-future articles.
Like in `populateState` function, here is the only option.
This PR analogously postpones to `getItems`.

- `getItems` function just uses `authorise` function to detect if user can edit articles.
This PR does all the previously posponed jobs: authorises, here at article level, and eventually filters out articles that are not published or are expired or future, if user hasn't 'edit' or 'edit.own' permissions.
To filter out articles that the user must not see, we remove them from the articles collection (the `$items` array) during usual `foreach` loop, so the successive code will just work on permitted/authorised articles.

3. featured model:
- `populateState` function authorises at component level and eventually filters only published articles.
Just like in articles model function this PS pospones to `getItems`.

### Testing Instructions
Create some **featured articles in different categories** (but "Uncategorised" would be enought); articles should have different publishing status: _published, unpublished, published-expired, unpublished-expired, published-in-future, unpublished-in-future_.

Create a **'Featured' menuitem**.

Set **permissions** in 'Uncategorised' category to allow 'Registered' group to 'edit' and 'edit.own'.

Create a user in registered group and login in frontend.


### Expected result
All the unpublished, expired and in-future articles in "Uncategorised" category should be displayed (as user has permissions to 'edit' and 'edit-own').

### Actual result
All the unpublished, expired and in-future articles in "Uncategorised" category are NOT displayed.


### Notes

1. I have not evaluated the performance impact of `foreach` loop in articles model `getItems` function that removes unwanted articles.

2. Should the `'filter.published'` settings be managed? How?
